### ISSUE TITLE
Fatal Error with WooCommerce due to too strict type contraint

### ISSUE THREAD

**Issue by BenceSzalai** [crstauf/query-monitor-extend#88]
Hi!

Query Monitor causes a Fatal error as:

```
PHP Fatal error: Uncaught TypeError: QMX_Collector_Image_Sizes::filter__wp_get_attachment_image_src(): Argument #2 ($attachment_id) must be of type int, string given, called in /var/www/html/wp-includes/class-wp-hook.php on line 326 and defined in /var/www/html/wp-content/plugins/query-monitor-extend/image-sizes/qmx-image-sizes-collector.php:148
```

The reason is that WooCommerce uses a default image when a product does not have one, and this image is the result of a `get_option( 'woocommerce_placeholder_image', 0 )` call which returns the image ID as a string which is passed as the first parameter to a subsequent `wp_get_attachment_image()` call (see the source of `wc_placeholder_img()` in `wp-content/plugins/woocommerce/includes/wc-product-functions.php` if interested).

Please remove the `int` type constraint from the 2nd argument of `QMX_Collector_Image_Sizes::filter__wp_get_attachment_image_src( $image, int $attachment_id, $size )` ([here](https://github.com/crstauf/query-monitor-extend/blob/master/image-sizes/qmx-image-sizes-collector.php#L148)). On one hand this argument is not used at all in the given function, so the type should not matter, and generally speaking any constraint that can cause such issues should be removed from Query Monitor. On the other hand even if the attachment ID was supposed to be an integer always (which I guess is not the case for many plugins just because how relaxed PHP and WordPress is regarding IDs as integers or strings) a Fatal Error is not the right way to handle such cases.

OR let me know if PR is welcome.

**Comment by crstauf** [comment]
Duplicate of #77.

**Comment by crstauf** [comment]
Internal debate occurring right now:

_While it is true that the function does not use the `$attachment_id` variable, and as such a strict type is not necessary, it is also problematic that other plugins use the incorrect type: we should all work together to provide a robust platform, especially a piece of software that is owned by the maker of WordPress._

🤔

**Comment by BenceSzalai** [comment]
Sorry, I didn't find that issue.

While I agree with the sentiment, the primary purpose of software is to work, and I assume anyone can honestly admit putting that `int` type-hint in the code won't have any affect on WooCommerce.

There is a long history of software having to adapt to mistakes in other sw and even hw components. I know it is not ideal, but at the end of the day SW should work.

I believe deliberately letting websites break just to send a signal to Automattic (do I read it correctly?) is a mistake for two reasons: it causes problem for users, not those who are responsible for the issue, and also there is no technical reason for the code to enforce this type.

I don't want to sound harsh, and please hear me out on not suggesting it is the same in scale, but I think it is similar to terrorism: knowingly causing harm to independent, "innocent" bystanders to emphasis a point.

**Comment by crstauf** [comment]
Yep, I have reached the same conclusion: I'm begrudgingly removing the type constraint.

**Comment by crstauf** [comment]
Next published version will remove the type constraint.

(Resolved by a41cb14a1899bea5582e3d50efb61f1095913635.)
